src/audio/host-zephyr.c

```diff
@@ -676,44 +591,80 @@ static struct comp_dev *host_new(const struct comp_driver *drv,
 	if (!hd->dma) {
 		comp_err(dev, "host_new(): dma_get() returned NULL");
 		rfree(hd);
-		rfree(dev);
-		return NULL;
+		return -ENODEV;
 	}
 
 	/* init buffer elems */
 	dma_sg_init(&hd->config.elem_array);
 	dma_sg_init(&hd->host.elem_array);
 	dma_sg_init(&hd->local.elem_array);
 
-	ipc_build_stream_posn(&hd->posn, SOF_IPC_STREAM_POSITION, dev->ipc_config.id);
+	ipc_build_stream_posn(&hd->posn, SOF_IPC_STREAM_POSITION, config_id);
 
 	hd->msg = ipc_msg_init(hd->posn.rhdr.hdr.cmd, sizeof(hd->posn));
 	if (!hd->msg) {
 		comp_err(dev, "host_new(): ipc_msg_init failed");
 		dma_put(hd->dma);
 		rfree(hd);
+		return -ENOMEM;
+	}
+	hd->chan = NULL;
+	hd->copy_type = COMP_COPY_NORMAL;
+
+	return 0;
+}
+
+static struct comp_dev *host_new(const struct comp_driver *drv,
+				 const struct comp_ipc_config *config,
+				 const void *spec)
+{
+	struct comp_dev *dev;
+	struct host_data *hd;
+	const struct ipc_config_host *ipc_host = spec;
+	int ret;
+
+	comp_cl_dbg(&comp_host, "host_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	hd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*hd));
+	if (!hd) {
 		rfree(dev);
 		return NULL;
 	}
+	comp_set_drvdata(dev, hd);
+
+	ret = host_zephyr_new(hd, dev, ipc_host, dev->ipc_config.id);
+	if (ret) {
+		rfree(hd);
+		rfree(dev);
+		comp_err(dev, "host: host zephyr new dma failed with exit");
+		return NULL;
+	}
 
-	hd->chan = NULL;
-	hd->copy_type = COMP_COPY_NORMAL;
 	dev->state = COMP_STATE_READY;
 
 	return dev;
 }
 
-static void host_free(struct comp_dev *dev)
+void host_zephyr_free(struct host_data *hd)
 {
-	struct host_data *hd = comp_get_drvdata(dev);
-
-	comp_dbg(dev, "host_free()");
-
 	dma_put(hd->dma);
 
 	ipc_msg_free(hd->msg);
 	dma_sg_free(&hd->config.elem_array);
 	rfree(hd);
+}
+
+static void host_free(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "host_free()");
+	host_zephyr_free(hd);
```

Comment:
hd is allocated in host_new() so it should be freed in host_free(0 to keep in balanced and easy to follow. This mean you will have to free it in copier_free() as well

Reply:
copier free have below code:
		if (cd->endpoint[i]->ipc_config.type == SOF_COMP_HOST && !cd->ipc_gtw) {
			host_zephyr_free(cd->hd);
			rfree(cd->endpoint[i]);
		} 

host_zephyr_free will free hd, is that ok?